### PR TITLE
Saving and loading state hooks

### PR DESCRIPTION
**Pull request by patrickvonplaten**
With this design we would have the necessary freedom to save & load models as wished in `diffusers`. 

Possible solution to: https://github.com/huggingface/accelerate/issues/976
The idea would be for loading:
- loop over the `models` and if a certain type of model, e.g. UNet, we call `from_pretrained(...)` and remove it from the list so it's not called after.

For saving:
- loop over the passed `self._models` and passed `weights` and if `self_models` if of certain type (e.g. UNet) we call `save_pretrained(...)` and remove only from the weights list so it's not called after. 

Design is heavily inspired by the hooks in: https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#Module

If this is ok, happy to write tests and better docs and try it out. 

cc @sgugger @pcuenca

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -18,11 +18,14 @@
...
 import warnings
+from collections import OrderedDict
 from contextlib import contextmanager
 from functools import wraps
-from typing import List, Optional, Union
+from typing import List, Optional, Union, Callable
 
 import torch
+import torch.utils.hooks as hooks
+from ...utils.hooks import RemovableHandle
```
Comment (sgugger) [inline src/accelerate/accelerator.py:28]: I don't think this object exists in accelerate.

**Review by sgugger — COMMENTED** [review]
I'd be interested to see how it solves your problem in practice: the existing code to load/save is still executed so I'm not sure how this actually helps. I can review more the design afterward (though I have no objection right now, just nits).

**Comment by patrickvonplaten** [comment]
The use case would then look like this in diffusers: https://github.com/huggingface/diffusers/pull/2048/files . Because we're doing loading and saving directly in the hooks, the weights and models have to be popped from the respective lists so that nothing is loaded / saved twice. 

Think there are also use cases where the model doesn't need to be popped, e.g. if someone wants to save the config for transformers in addition to the checkpoint the hooks could just save and load the config but then let accelerate do the saving / loading of the weights. 

Think this design is quite minimal for accelerate. Wdyt? 
Happy to add a test to accelerate that shows how one could use the hooks to save a config of a transformer model in addition to saving/loading the weights.

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -1613,6 +1619,11 @@ def save(self, obj, f):
         """
         save(obj, f)
 
+    def register_save_state_pre_hook(self, hook: Callable[..., None]) -> hooks.RemovableHandle:
+        handle = hooks.RemovableHandle(self._save_model_state_pre_hook)
+        self._save_model_state_pre_hook[handle.id] = hook
+        return handle
+
```
Comment (pacman100) [inline src/accelerate/accelerator.py:1626]: Can we combine `register_save_state_pre_hook` and `register_load_state_pre_hook` using a bool parameter `load_state`

```suggestion
    def register_pre_hook(self, hook: Callable[..., None], load_state: bool) -> hooks.RemovableHandle:
        if load_state:
            pre_hook = self._load_model_state_pre_hook
        else:
            pre_hook = self._save_model_state_pre_hook
        handle = hooks.RemovableHandle(pre_hook)
        pre_hook[handle.id] = hook
        return handle

```

**Review by pacman100 — COMMENTED** [review]
Great way to enable custom saving and loading, pretty neat! ✨

Left a comment.

File: src/accelerate/accelerator.py (hunk 2 above)
Comment (muellerzr) [reply src/accelerate/accelerator.py:1626]: @pacman100 WDYT of instead mimicking how we have `prepare` and have a `register_state_pre_hook` func that calls internally `register_save_state_pre_hook` or `register_load_state_pre_hook`? I also think the explicit naming on `save` vs `load` pre-hook is important here for users to understand what it does, but have `save_state_pre_hook` for convenience
potentially.

File: tests/test_accelerator.py [hunk 3]
```diff
@@ -49,3 +61,83 @@ def test_free_memory_dereferences_prepared_components(self):
         self.assertTrue(len(accelerator._schedulers) == 0)
         self.assertTrue(len(accelerator._dataloaders) == 0)
         AcceleratorState._reset_state()
+
+    def test_save_load_model(self):
```
Comment (patrickvonplaten) [inline tests/test_accelerator.py:65]: This test is a bit unrelated to the PR, but think this is a useful test for the accelerator class in general.

**Comment by patrickvonplaten** [comment]
Pretty sure test failures are unrelated. Would be great to get another review here :-)

File: src/accelerate/accelerator.py [hunk 4]
```diff
@@ -1613,6 +1619,38 @@ def save(self, obj, f):
...
+        `hook(models: List[torch.nn.Module], weights: List[Dict[str, torch.Tensor]], input_dir: str) -> None`
+
+        The `models` argument are the models as saved in the accelerator state under `accelerator._models`, `weigths`
+        argument are the state dicts of the `models`, and the `input_dir` argument is the `input_dir` argument passed
+        to [`Accelerator.load_state`].
+
+        <Tip>
+
+        Should only be used in conjunction with [`Accelerator.register_save_state_pre_hook`]. Can be useful to save
```
Comment (sgugger) [inline src/accelerate/accelerator.py:1640]:
```suggestion
        Should only be used in conjunction with [`Accelerator.register_load_state_pre_hook`]. Can be useful to save
```

**Review by sgugger — APPROVED** [review]
Nice, thanks a lot for iterating on this!

**Comment by muellerzr** [comment]
@patrickvonplaten only was now able to look at this today, I can confirm that your PR does indeed break that test, it works prior to this and is only on the CPU when it fails. If you have some time please try to look into it otherwise I'll look into this later today. cc @sgugger

**Pull request merged**